common/djangoapps/student/models.py

```diff
@@ -331,6 +362,79 @@ def user_profile_pre_save_callback(sender, **kwargs):
     if user_profile.requires_parental_consent() and user_profile.has_profile_image:
         user_profile.profile_image_uploaded_at = None
 
+    # Cache "old" field values on the model instance so that they can be
+    # retrieved in the post_save callback when we emit an event with new and
+    # old field values.
+    user_profile._changed_fields = get_changed_fields_dict(user_profile, sender)
+
+@receiver(post_save, sender=UserProfile)
+def user_profile_post_save_callback(sender, **kwargs):
+    """
+    Emit analytics events after saving the UserProfile.
+    """
+    user_profile = kwargs['instance']
+    # pylint: disable=protected-access
+    emit_field_changed_events(
+        user_profile, user_profile.user, USER_SETTINGS_CHANGED_EVENT_NAME, sender._meta.db_table, ['meta']
+    )
+
+
+# @receiver(pre_save, sender=User)
+# def user_pre_save_callback(sender, **kwargs):
+#     """
+#     Capture old fields on the user instance before save and cache them as a
+#     private field on the current model for use in the post_save callback.
+#     """
+#     user = kwargs['instance']
+#     user._changed_fields = get_changed_fields_dict(user, sender)
+
+
+# @receiver(post_save, sender=User)
+# def user_post_save_callback(sender, **kwargs):
+#     """
+#     Emit analytics events after saving the User.
+#     """
+#     user = kwargs['instance']
+#     # pylint: disable=protected-access
+#     emit_field_changed_events(
+#         user, user, USER_SETTINGS_CHANGED_EVENT_NAME, sender._meta.db_table, hidden_fields=['password']
+#     )
+
+def emit_field_changed_events(instance, user, event_name, db_table, excluded_fields=None, hidden_fields=None):
```

Comment:
Could you move this to be a helper as it doesn't appear to have any special knowledge of the student models. In particular, I'm wondering how I can reuse this for UserPreference which lives in the user_api app.